RFC #30: React.forwardRef implementation

bvaughn opened the pull request:
Add new API, `React.forwardRef`, implemented as a new element type. Usage is e.g.:

```js
const ThemeContext = React.createContext("light");

// Example HOC
function withTheme(ThemedComponent) {
  function ThemeContextInjector(props) {
    return (
      <ThemeContext.Consumer>
        {value => (
          <ThemedComponent {...props} ref={props.forwardedRef} theme={value} />
        )}
      </ThemeContext.Consumer>
    );
  }

  // Forward refs through to the inner, "themed" component:
  return React.forwardRef((props, ref) => (
    <ThemeContextInjector {...props} forwardedRef={ref} />
  ));
}

const ThemedButton = withTheme(Button);
const buttonRef = React.createRef();

// buttonRef.value will point to ThemedButton, rather than ThemeContextInjector
<ThemedButton ref={buttonRef} />;
```

Relates to reactjs/rfcs/pull/30

TrySound commented on packages/react/src/useRef.js:
```diff
@@ -0,0 +1,25 @@
+) {
+  warning(
+    typeof renderProp === 'function',
+    'useRef requires a render prop but was given %s.',
```
"Render prop" in warning may confuse users. Maybe "render function"?

bvaughn replied:
Sure, that's reasonable 👍

jquense commented:
can I just say that It's really cool how async mode as inadvertently tipped the scales on all these long running TBD issues like context and refs :)

👏 Awesome work Brian! 🙇

bvaughn commented:
Agreed!

Thanks, Jason. 😄

sebmarkbage commented:
@jquense (That is true that it’s a forcing function but a little bit of the credit goes to us having a 3 times larger and ramped up amazing team now. ❤️)

jquense commented:
❤️ ya go team! y'all are crushing it

gaearon commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -166,6 +167,16 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
     return workInProgress.child;
   }
 
+  function updateUseRef(current, workInProgress) {
```
Wow why are these things so easy with Fiber

bvaughn replied:
I know! 😁

gaearon commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -166,6 +167,16 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
   }
 
+  function updateUseRef(current, workInProgress) {
+    const nextChildren = workInProgress.type.renderProp(
```
Might want to validate before calling to get a nicer error message. See a similar check before context. I think Seb told me once it doesn’t affect perf because engine has to check it’s a function regardless and will likely remember that.

bvaughn replied:
I already validate in `packages/react/src/useRef.js` when the prop is passed initially.

aweary replied:
The validation in `useRef.js` is just a warning though, maybe use an invariant and throw early instead?

gaearon replied:
Hmm maybe we should change context to work the same way then.

I don’t think we should make it ever invariant during creation. Makes inlining harder and will likely cause duplicate checks.

aweary replied:
Is the intention to rely on the warning to explain the error that would occur here when `renderProp` isn't a function? I understand not throwing during creation, but an explicit error seems better to me.

aweary replied:
I think warning during creation is a good idea, but I also think there should be an explicit error with a component stack trace when `updateUseRef` inevitably throws.

bvaughn replied:
Is 166d236 what you had in mind? The component stack will only be included in DEV mode.

This feels a little weird to me, checking both places.

sebmarkbage commented:
Warning > Invariant

bvaughn commented:
I assume @sebmarkbage's comment is directed towards @aweary?

This PR is using `warning` 😄

aweary commented:
(moving this conversation out of the collapsed comment thread)

> Is 166d236 what you had in mind? The component stack will only be included in DEV mode.
> This feels a little weird to me, checking both places.

@bvaughn having a warning and invariant seems fine to me. There's precedence for it as well; you get a warning at creation time for invalid element types, and an error during render. Maybe the component stack could go in the warning in that case?

bvaughn commented:
> Maybe the component stack could go in the warning in that case?

There's no component stack for the warning, right?

aweary commented:
Ah, right, that's true 🙂I don't think the component stack critical. If it can be included in the invariant, that's cool (errors get stripped in production anyway, right? so it wouldn't matter if it's DEV-only) but if not that's probably fine too.

acdlite commented:
It occurs to me that we might be able to avoid a new Fiber type by storing the `ref` attribute as a non-enumerable symbol property. Then, `useRef` could look like this:

```js
function useRef(props, ...rest) {
  const ref = props[REF_SYMBOL];  
  return component => {
    return component(props, ref, ...rest);
  };
}
```

It's possible I'm overlooking something. Also not sure if this is better or worse.

bvaughn commented:
I dig that suggestion 👍

bvaughn commented:
Renamed `useRef` to `forwardRef`

sebmarkbage commented on packages/react/src/forwardRef.js:
```diff
@@ -0,0 +1,25 @@
+export default function forwardRef<Props, ElementType: React$ElementType>(
+  renderProp: (props: Props, ref: React$ElementRef<ElementType>) => React$Node,
+) {
+  warning(
```
`if (__DEV__)` around it, right?

bvaughn replied:
`warning` is an empty function if not DEV, so I didn't think that was necessary. But sure!

sebmarkbage commented on packages/react/src/forwardRef.js:
```diff
@@ -0,0 +1,25 @@
+import warning from 'fbjs/lib/warning';
+
+export default function forwardRef<Props, ElementType: React$ElementType>(
+  renderProp: (props: Props, ref: React$ElementRef<ElementType>) => React$Node,
```
This is not really a render prop since it's not passed at render time nor a prop. It's more like, just `renderFn` or `render` maybe?

bvaughn replied:
Good suggestion. I like `renderFn`

acdlite commented on packages/shared/ReactSymbols.js:
```diff
@@ -36,6 +36,9 @@ export const REACT_CONTEXT_TYPE = hasSymbol
   ? Symbol.for('react.async_mode')
   : 0xeacf;
+export const REACT_USE_REF_TYPE = hasSymbol
+  ? Symbol.for('react.use_ref')
```
Let's rename these too?

bvaughn replied:
Yup. I kind of halted my renaming in the middle because of the `ReactChildFiber` suggestion I'm trying out.

bvaughn commented:
I've updated the approach to call through in `ReactChildFiber` and avoid creating the fiber. I'm not super familiar with the code I changed though, so I may have overlooked a codepath.

sebmarkbage commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -1128,6 +1149,32 @@ function ChildReconciler(shouldTrackSideEffects) {
       child = child.sibling;
```
The element.type won't line up on line 1125 which means that it'll kill the state of the subtree every time this rerenders.

sebmarkbage commented on packages/react-reconciler/src/ReactChildFiber.js:
```diff
@@ -1128,6 +1149,32 @@ function ChildReconciler(shouldTrackSideEffects) {
+        return null;
+      }
+
+      return reconcileSingleElement(
```
We've tried pretty hard to avoid recursive algorithms so that we could try to inline these into hot loops instead. Would prefer a while loop.

bvaughn commented:
@sebmarkbage and I talked offline about this and decided to revert to my initial approach for now (the new fiber/type). We will address the no-fiber optimization in a follow-up; it also impacts stateless functional components, and will be a bit tricky to get right.

bvaughn commented:
I've reverted the changes to `ReactChildFiber` and gone back to the original function type, but with the name `React.forwardRef` instead of `React.useRef`.

bvaughn commented:
SSR partial renderer has been updated. Working on shallow renderer now.

bvaughn commented:
I don't know if it really makes sense to update `ReactShallowRenderer` to support this new component type. (This renderer doesn't support the new context API either.)

gaearon commented on packages/shared/ReactSymbols.js:
```diff
@@ -36,6 +36,9 @@ export const REACT_CONTEXT_TYPE = hasSymbol
   ? Symbol.for('react.async_mode')
   : 0xeacf;
+export const REACT_FORWARD_REF_TYPE = hasSymbol
+  ? Symbol.for('react.use_ref')
```
Needs rename

gaearon commented on packages/react-reconciler/src/ReactFiberBeginWork.js:
```diff
@@ -153,6 +154,16 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
   }
 
+  function updateForwardRef(current, workInProgress) {
+    const nextChildren = workInProgress.type.render(
```
Let's read `render` first and then call it. Then the fact that it's called with `type` as `this` doesn't become an unintenional part of the public API.

bvaughn replied:
Good call.

gaearon commented on packages/react/src/__tests__/forwardRef-test.internal.js:
```diff
@@ -0,0 +1,252 @@
+  beforeEach(() => {
+    jest.resetModules();
+    ReactFeatureFlags = require('shared/ReactFeatureFlags');
+    ReactFeatureFlags.debugRenderPhaseSideEffectsForStrictMode = false;
```
Is this necessary for this test? If not you can remove the internal file name suffix.

bvaughn replied:
Without it, the number of yields vary between dev and prod :(

gaearon replied:
Got it. Okay.

gaearon commented on packages/react/src/__tests__/forwardRef-test.internal.js:
```diff
@@ -0,0 +1,252 @@
+
+    ReactNoop.render(<RefForwardingComponent ref={ref} setRefOnDiv={true} />);
+    ReactNoop.flush();
+    expect(ref.value.type).toBe('div');
```
Needs rebase? Current.

timneutkens commented on packages/react-reconciler/src/ReactFiberCommitWork.js:
```diff
@@ -315,7 +316,19 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
+            warning(
+              false,
+              'Unexpected ref object provided for %s. ' +
+                'Use either a ref-setter function or Reacte.createRef().%s',
```
@bvaughn `Reacte` is a typo right 🤔

gaearon replied:
Haha yeah. I think that was fixed in master but would appreciate a double check

bvaughn replied:
😁 Yeah.
